From ee17d76440a5342632821e5b30b0fe7444d47cd8 Mon Sep 17 00:00:00 2001
From: Shilin Victor <chrono.monochrome@gmail.com>
Date: Mon, 20 Aug 2018 22:57:43 +0300
Subject: [PATCH 036/296] fix service could not be started correctly when
 app.thread == null

there is not thread object attached to application when starting service,
will not start process, so make start service failed.

startActivity, broadcast or  provider all have condition:
 if app == null || app.thread == null,  that will start process.
but startService only if app == null, lack of  condition app.thread == null,  that will start process.
---
 services/core/java/com/android/server/am/ActiveServices.java | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/services/core/java/com/android/server/am/ActiveServices.java b/services/core/java/com/android/server/am/ActiveServices.java
index f8b0684f8ed..f6d91adf69e 100644
--- a/services/core/java/com/android/server/am/ActiveServices.java
+++ b/services/core/java/com/android/server/am/ActiveServices.java
@@ -2146,7 +2146,7 @@ public final class ActiveServices {
 
         // Not running -- get it started, and enqueue this service record
         // to be executed when the app comes up.
-        if (app == null && !permissionsReviewRequired) {
+        if ((app == null || app.thread == null) && !permissionsReviewRequired) {
             if ((app=mAm.startProcessLocked(procName, r.appInfo, true, intentFlags,
                     hostingType, r.name, false, isolated, false)) == null) {
                 String msg = "Unable to launch app "
-- 
2.11.0

